doc/meek-client.1
'\" t
.\"     Title: meek-client
.\"    Author: [FIXME: author] [see http://docbook.sf.net/el/author]
.\" Generator: DocBook XSL Stylesheets v1.79.1 <http://docbook.sf.net/>
.\"      Date: 09/03/2021
.\"    Manual: \ \&
.\"    Source: \ \&
.\"  Language: English
.\"
.TH "MEEK\-CLIENT" "1" "09/03/2021" "\ \&" "\ \&"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
meek-client \- The meek client transport plugin
.SH "SYNOPSIS"
.sp
\fBmeek\-client\fR [\fIOPTIONS\fR]
.SH "DESCRIPTION"
.sp
meek\-client is a transport plugin for Tor that encodes a stream as a sequence of HTTP requests and responses\&. It has a \fBurl\fR option to control what destination server requests are directed to, and a \fBfront\fR option for domain name camouflage: The domain name in the URL is replaced by the front domain before the request is made, but the Host header inside the HTTP request still points to the original domain\&. The idea is to front through a domain that is not blocked to a domain that is blocked\&.
.SH "CONFIGURATION"
.sp
Configuration for meek\-client usually happens in a torrc file\&.
.sp
Per\-bridge options are configured with SOCKS args (key=value pairs in a Bridge line)\&. The possible SOCKS args are:
.PP
\fBurl\fR=\fIURL\fR (required)
.RS 4
The URL of a meek\-server instance\&. The domain name component will typically be hidden by the value in the
\fBfront\fR
arg\&.
.RE
.PP
\fBfront\fR=\fIDOMAIN\fR
.RS 4
Front domain name\&. If provided, this domain name will replace the domain name of
\fBurl\fR
in the DNS request and TLS SNI field\&. The URL\(cqs true domain name will still appear in the Host header of HTTP requests\&.
.RE
.PP
\fButls\fR=\fICLIENTHELLOID\fR
.RS 4
Use the
uTLS library
with the named TLS fingerprint for TLS camouflage\&. This arg is incompatible with the
\fB\-\-helper\fR
command line option\&. The possible values of
\fICLIENTHELLOID\fR
are:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
HelloRandomizedALPN
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
HelloRandomizedNoALPN
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
HelloFirefox_55
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
HelloFirefox_56
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
HelloFirefox_63
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
HelloFirefox_65
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
HelloFirefox_Auto = HelloFirefox_65
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
HelloChrome_58
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
HelloChrome_62
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
HelloChrome_70
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
HelloChrome_72
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
HelloChrome_83
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
HelloChrome_Auto = HelloChrome_83
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
HelloIOS_11_1
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
HelloIOS_12_1
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
HelloIOS_Auto = HelloIOS_12_1
.RE
.sp
As a special case, the values "none" and "HelloGolang" are recognized as aliases for omitting the
\fButls\fR
SOCKS arg; i\&.e\&., use native Go TLS\&.
.RE
.sp
For backward compatibility, each SOCKS arg also has an equivalent command line option\&. For example, this configuration using SOCKS args:
.sp
.if n \{\
.RS 4
.\}
.nf
Bridge meek 0\&.0\&.2\&.0:1 url=https://forbidden\&.example/ front=allowed\&.example
ClientTransportPlugin meek exec \&./meek\-client
.fi
.if n \{\
.RE
.\}
.sp
is the same as this one using command line options:
.sp
.if n \{\
.RS 4
.\}
.nf
Bridge meek 0\&.0\&.2\&.0:1
ClientTransportPlugin meek exec \&./meek\-client \-\-url=https://forbidden\&.example/ \-\-front=allowed\&.example
.fi
.if n \{\
.RE
.\}
.sp
SOCKS args are preferred over command line options because they allow you to have multiple Bridge lines with different settings\&. If a SOCKS arg and a command line option are both given for the same setting, the SOCKS arg takes precedence\&.
.sp
The global \fB\-\-helper\fR option prevents meek\-client from doing any network operations itself\&. Rather, it will send all requests through a browser extension, which must be set up separately\&.
.sp
A global proxy (applies to all Bridge lines) can be configured using special torrc options:
.sp
.if n \{\
.RS 4
.\}
.nf
Socks4Proxy localhost:1080
Socks5Proxy localhost:1080
Socks5ProxyUsername username
Socks5ProxyPassword password
HTTPSProxy localhost:8080
HTTPSProxyAuthenticator username:password
.fi
.if n \{\
.RE
.\}
.sp
or, equivalently, using the \fB\-\-proxy\fR command line option\&. The command line option takes precedence\&.
.sp
When the \fB\-\-helper\fR option is used, you can use proxies of type http, socks4a, or socks5, but you cannot use a username or password with the proxy\&. Without \fB\-\-helper\fR, you can use proxies of type http, https, or socks5, and you can optionally use a username and password\&.
.SH "OPTIONS"
.PP
\fB\-\-front\fR=\fIDOMAIN\fR
.RS 4
Front domain name\&. Prefer using the
\fBfront\fR
SOCKS arg on a bridge line over using this command line option\&.
.RE
.PP
\fB\-\-helper\fR=\fIADDRESS\fR
.RS 4
Address of HTTP helper browser extension\&. For example,
\fB\-\-helper=127\&.0\&.0\&.1:7000\fR\&.
.RE
.PP
\fB\-\-proxy\fR=\fIURL\fR
.RS 4
URL of upstream proxy\&. For example,
\fB\-\-proxy=http://localhost:8080/\fR,
\fB\-\-proxy=socks4a://localhost:1080\fR, or
\fB\-\-proxy=socks5://localhost:1080\fR\&. Can also be configured using the
\fBHTTPSProxy\fR,
\fBSocks4Proxy\fR, or
\fBSocks5Proxy\fR
options in a torrc file\&.
.RE
.PP
\fB\-\-log\fR=\fIFILENAME\fR
.RS 4
Name of a file to write log messages to (default stderr)\&.
.RE
.PP
\fB\-\-url\fR=\fIURL\fR
.RS 4
URL to correspond with\&. Prefer using the
\fBurl\fR
SOCKS arg on a bridge line over using this command line option\&.
.RE
.PP
\fB\-\-utls\fR=\fICLIENTHELLOID\fR
.RS 4
Use uTLS with the given TLS fingerprint for TLS camouflage\&. This option is incompatible with
\fB\-\-helper\fR\&. Prefer using the
\fButls\fR
SOCKS arg over using this command line option\&.
.RE
.PP
\fB\-h\fR, \fB\-\-help\fR
.RS 4
Display a help message and exit\&.
.RE
.SH "SEE ALSO"
.sp
\fBhttps://trac\&.torproject\&.org/projects/tor/wiki/doc/meek\fR
.SH "BUGS"
.sp
Please report at \fBhttps://trac\&.torproject\&.org/projects/tor\fR\&.


meek-client/meek-client.go
// meek-client is the client transport plugin for the meek pluggable transport.
//
// Sample usage in torrc:
// 	Bridge meek 0.0.2.0:1 url=https://forbidden.example/ front=allowed.example
// 	ClientTransportPlugin meek exec ./meek-client
// The transport ignores the bridge address 0.0.2.0:1 and instead connects to
// the URL given by url=. When front= is given, the domain in the URL is
// replaced by the front domain for the purpose of the DNS lookup, TCP
// connection, and TLS SNI, but the HTTP Host header in the request will be the
// one in url=.
//
// Most user configuration can happen either through SOCKS args (i.e., args on a
// Bridge line) or through command line options. SOCKS args take precedence
// per-connection over command line options. For example, this configuration
// using SOCKS args:
// 	Bridge meek 0.0.2.0:1 url=https://forbidden.example/ front=allowed.example
// 	ClientTransportPlugin meek exec ./meek-client
// is the same as this one using command line options:
// 	Bridge meek 0.0.2.0:1
// 	ClientTransportPlugin meek exec ./meek-client --url=https://forbidden.example/ --front=allowed.example
// The command-line configuration interface is for compatibility with tor 0.2.4
// and older, which doesn't support parameters on Bridge lines.
//
// The --helper option prevents this program from doing any network operations
// itself. Rather, it will send all requests through a browser extension that
// makes HTTP requests.
package main

import (
	"flag"
	"fmt"
	"io"
	"io/ioutil"
	"log"
	"net"
	"net/http"
	"net/url"
	"os"
	"os/signal"
	"sync"
	"syscall"
	"time"

	"git.torproject.org/pluggable-transports/goptlib.git"
	"github.com/xtaci/kcp-go/v5"
	"github.com/xtaci/smux"
)

const (
	ptMethodName = "meek"
	// Safety limits on interaction with the HTTP helper.
	maxHelperResponseLength = 10000000
	helperReadTimeout       = 60 * time.Second
	helperWriteTimeout      = 2 * time.Second

	// How long before timing out connections at the inner KCP layer.
	smuxIdleTimeout = 30 * time.Minute
)

// We use this RoundTripper to make all our requests when neither --helper nor
// utls is in effect. We use the defaults, except we take control of the Proxy
// setting (notably, disabling the default ProxyFromEnvironment).
var httpRoundTripper *http.Transport = http.DefaultTransport.(*http.Transport)

// We use this RoundTripper when --helper is in effect.
var helperRoundTripper = &HelperRoundTripper{
	ReadTimeout:  helperReadTimeout,
	WriteTimeout: helperWriteTimeout,
}

// Store for command line options.
var options struct {
	URL       string
	Front     string
	ProxyURL  *url.URL
	UseHelper bool
	UTLSName  string
}

// urlAddr is a net.Addr representation of a url.URL.
type urlAddr struct{ *url.URL }

func (addr urlAddr) Network() string { return "url" }
func (addr urlAddr) String() string  { return addr.URL.String() }

// RequestInfo encapsulates all the configuration used for a request–response
// roundtrip, including variables that may come from SOCKS args or from the
// command line.
type RequestInfo struct {
	// The URL to request.
	URL *url.URL
	// The Host header to put in the HTTP request (optional and may be
	// different from the host name in URL).
	Host string
	// The RoundTripper to use to send requests. This may vary depending on
	// the value of global options like --helper.
	RoundTripper http.RoundTripper
}

func (info *RequestInfo) Poll(out io.Reader) (in io.ReadCloser, err error) {
	req, err := http.NewRequest("POST", info.URL.String(), out)
	// Prevent Content-Type sniffing by net/http and middleboxes.
	req.Header.Set("Content-Type", "application/octet-stream")
	if err != nil {
		return nil, err
	}
	if info.Host != "" {
		req.Host = info.Host
	}
	if err != nil {
		return nil, err
	}
	resp, err := info.RoundTripper.RoundTrip(req)
	if err == nil && resp.StatusCode != http.StatusOK {
		err = fmt.Errorf("status code %d", resp.StatusCode)
	}
	if err != nil {
		return nil, err
	}
	return resp.Body, nil
}

// Callback for new SOCKS requests.
func handleSOCKS(socks *pt.SocksConn) error {
	var info RequestInfo

	// First check url= SOCKS arg, then --url option.
	urlArg, ok := socks.Req.Args.Get("url")
	if ok {
	} else if options.URL != "" {
		urlArg = options.URL
	} else {
		return fmt.Errorf("no URL for SOCKS request")
	}
	var err error
	info.URL, err = url.Parse(urlArg)
	if err != nil {
		return err
	}

	// First check front= SOCKS arg, then --front option.
	front, ok := socks.Req.Args.Get("front")
	if ok {
	} else if options.Front != "" {
		front = options.Front
		ok = true
	}
	if ok {
		info.Host = info.URL.Host
		info.URL.Host = front
	}

	// First check utls= SOCKS arg, then --utls option.
	utlsName, utlsOK := socks.Req.Args.Get("utls")
	if utlsOK {
	} else if options.UTLSName != "" {
		utlsName = options.UTLSName
		utlsOK = true
	}

	// First we check --helper: if it was specified, then we always use the
	// helper, and utls is disallowed. Otherwise, we use utls if requested;
	// or else fall back to native net/http.
	if options.UseHelper {
		if utlsOK {
			return fmt.Errorf("cannot use utls with --helper")
		}
		info.RoundTripper = helperRoundTripper
	} else if utlsOK {
		info.RoundTripper, err = NewUTLSRoundTripper(utlsName, nil, options.ProxyURL)
		if err != nil {
			return err
		}
	} else {
		info.RoundTripper = httpRoundTripper
	}

	// Each SOCKS connection corresponds to a single KCP session with a
	// single smux stream inside it.
	pconn := NewPollingPacketConn(urlAddr{info.URL}, &info)
	defer pconn.Close()

	conn, err := kcp.NewConn2(pconn.RemoteAddr(), nil, 0, 0, pconn)
	if err != nil {
		return err
	}
	defer conn.Close()
	// Permit coalescing the payloads of consecutive sends.
	conn.SetStreamMode(true)
	// Disable the dynamic congestion window (limit only by the
	// maximum of local and remote static windows).
	conn.SetNoDelay(
		0, // default nodelay
		0, // default interval
		0, // default resend
		1, // nc=1 => congestion window off
	)

	smuxConfig := smux.DefaultConfig()
	smuxConfig.Version = 2
	smuxConfig.KeepAliveTimeout = smuxIdleTimeout
	sess, err := smux.Client(conn, smuxConfig)
	if err != nil {
		return err
	}
	defer sess.Close()

	stream, err := sess.OpenStream()
	if err != nil {
		return err
	}
	defer stream.Close()

	err = socks.Grant(&net.TCPAddr{IP: net.IPv4zero, Port: 0})
	if err != nil {
		return err
	}

	var wg sync.WaitGroup
	wg.Add(2)
	go func() {
		defer wg.Done()
		_, err := io.Copy(socks, stream)
		if err != nil {
			log.Printf("recv error: %v", err)
		}
		err = socks.Close()
		if err != nil {
			log.Printf("socks shutdown error: %v", err)
		}
	}()
	go func() {
		defer wg.Done()
		_, err := io.Copy(stream, socks)
		if err != nil {
			log.Printf("send error: %v", err)
		}
		err = stream.Close()
		if err != nil {
			log.Printf("stream close error: %v", err)
		}
	}()
	wg.Wait()

	return nil
}

func acceptSOCKS(ln *pt.SocksListener) error {
	defer ln.Close()
	for {
		conn, err := ln.AcceptSocks()
		if err != nil {
			log.Printf("error in AcceptSocks: %s", err)
			if e, ok := err.(net.Error); ok && e.Temporary() {
				continue
			}
			return err
		}
		go func() {
			defer conn.Close()
			err := handleSOCKS(conn)
			if err != nil {
				conn.Reject()
				log.Printf("error in handling request: %s", err)
			}
		}()
	}
	return nil
}

// Return an error if this proxy URL doesn't work with the rest of the
// configuration.
func checkProxyURL(u *url.URL) error {
	if !options.UseHelper {
		// Without the helper, we use net/http's built-in proxy support,
		// which allows "http", "https", and "socks5".
		// socks5 requires go1.9: https://golang.org/doc/go1.9#net/http
		// https requires go1.10: https://golang.org/doc/go1.10#net/http
		// If using an older version of Go, the proxy won't be bypassed;
		// you'll just get an error at connection time rather than
		// TOR_PT_PROXY time.
		switch u.Scheme {
		case "http", "https", "socks5":
		default:
			return fmt.Errorf("don't understand proxy URL scheme %q", u.Scheme)
		}
	} else {
		// With the helper we can use HTTP and SOCKS (because it is the
		// browser that does the proxying, not us).
		// For the HTTP proxy with the Firefox helper: versions of
		// Firefox before 32, and Tor Browser before 3.6.2, leak the
		// covert Host header in HTTP proxy CONNECT requests. Using an
		// HTTP proxy cannot provide effective obfuscation without such
		// a patched Firefox.
		// https://bugs.torproject.org/12146
		// https://gitweb.torproject.org/tor-browser.git/commit/?id=e08b91c78d919f66dd5161561ca1ad7bcec9a563
		// https://bugzilla.mozilla.org/show_bug.cgi?id=1017769
		// https://hg.mozilla.org/mozilla-central/rev/a1f6458800d4
		switch u.Scheme {
		case "http", "socks5", "socks4a":
		default:
			return fmt.Errorf("don't understand proxy URL scheme %q", u.Scheme)
		}
		if u.User != nil {
			return fmt.Errorf("a proxy URL with a username or password can't be used with --helper")
		}
	}
	return nil
}

func main() {
	var helperAddr string
	var logFilename string
	var proxy string
	var err error

	flag.StringVar(&options.Front, "front", "", "front domain name if no front= SOCKS arg")
	flag.StringVar(&helperAddr, "helper", "", "address of HTTP helper (browser extension)")
	flag.StringVar(&logFilename, "log", "", "name of log file")
	flag.StringVar(&proxy, "proxy", "", "proxy URL")
	flag.StringVar(&options.URL, "url", "", "URL to request if no url= SOCKS arg")
	flag.StringVar(&options.UTLSName, "utls", "", "uTLS Client Hello ID")
	flag.Parse()

	ptInfo, err := pt.ClientSetup(nil)
	if err != nil {
		log.Fatalf("error in ClientSetup: %s", err)
	}

	log.SetFlags(log.LstdFlags | log.LUTC)
	if logFilename != "" {
		f, err := os.OpenFile(logFilename, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0600)
		if err != nil {
			// If we fail to open the log, emit a message that will
			// appear in tor's log.
			pt.CmethodError(ptMethodName, fmt.Sprintf("error opening log file: %s", err))
			log.Fatalf("error opening log file: %s", err)
		}
		defer f.Close()
		log.SetOutput(f)
	}

	if helperAddr != "" {
		options.UseHelper = true
		helperRoundTripper.HelperAddr, err = net.ResolveTCPAddr("tcp", helperAddr)
		if err != nil {
			log.Fatalf("can't resolve helper address: %s", err)
		}
		log.Printf("using helper on %s", helperRoundTripper.HelperAddr)
	}

	if proxy != "" {
		options.ProxyURL, err = url.Parse(proxy)
		if err != nil {
			log.Fatalf("can't parse proxy URL: %s", err)
		}
	}

	// Disable the default ProxyFromEnvironment setting.
	// httpRoundTripper.Proxy is overridden below if options.ProxyURL is
	// set.
	httpRoundTripper.Proxy = nil

	// Command-line proxy overrides managed configuration.
	if options.ProxyURL == nil {
		options.ProxyURL = ptInfo.ProxyURL
	}
	// Check whether we support this kind of proxy.
	if options.ProxyURL != nil {
		err = checkProxyURL(options.ProxyURL)
		if err != nil {
			pt.ProxyError(err.Error())
			log.Fatal(fmt.Sprintf("proxy error: %s", err))
		}
		log.Printf("using proxy %s", options.ProxyURL.String())
		httpRoundTripper.Proxy = http.ProxyURL(options.ProxyURL)
		if options.UseHelper {
			err = helperRoundTripper.SetProxy(options.ProxyURL)
			if err != nil {
				pt.ProxyError(err.Error())
				log.Fatal(fmt.Sprintf("proxy error: %s", err))
			}
		}
		if ptInfo.ProxyURL != nil {
			pt.ProxyDone()
		}
	}

	listeners := make([]net.Listener, 0)
	for _, methodName := range ptInfo.MethodNames {
		switch methodName {
		case ptMethodName:
			ln, err := pt.ListenSocks("tcp", "127.0.0.1:0")
			if err != nil {
				pt.CmethodError(methodName, err.Error())
				break
			}
			go acceptSOCKS(ln)
			pt.Cmethod(methodName, ln.Version(), ln.Addr())
			log.Printf("listening on %s", ln.Addr())
			listeners = append(listeners, ln)
		default:
			pt.CmethodError(methodName, "no such method")
		}
	}
	pt.CmethodsDone()

	sigChan := make(chan os.Signal, 1)
	signal.Notify(sigChan, syscall.SIGTERM)

	if os.Getenv("TOR_PT_EXIT_ON_STDIN_CLOSE") == "1" {
		// This environment variable means we should treat EOF on stdin
		// just like SIGTERM: https://bugs.torproject.org/15435.
		go func() {
			io.Copy(ioutil.Discard, os.Stdin)
			log.Printf("synthesizing SIGTERM because of stdin close")
			sigChan <- syscall.SIGTERM
		}()
	}

	// Wait for a signal.
	sig := <-sigChan
	log.Printf("got signal %s", sig)

	for _, ln := range listeners {
		ln.Close()
	}

	log.Printf("done")
}
